proposals/2019-08-26-json-module.md

```diff
@@ -0,0 +1,99 @@
+
+* Bazel's [`struct`](https://docs.bazel.build/versions/master/skylark/lib/struct.html) type has a `to_json()` method
+  that can generate JSON documents. Issue [bazelbuild/bazel#7879](https://github.com/bazelbuild/bazel/issues/7879)
+  requests additional control over the behavior of this method (i.e. optional whitespace).
+
+* [bazelbuild/bazel#3732](https://github.com/bazelbuild/bazel/issues/3732) requests support for parsing JSON from
+  within a Bazel repository rule, for use with introspecting the state of external tools.
+
+* [`json_parser.bzl`](https://github.com/erickj/bazel_json) is a JSON parser implemented entirely within Starlark.
+  Limitations of the Starlark language prevent this parser from being recommended for production use.
+
+* [google/starlark-go#179](https://github.com/google/starlark-go/pull/179) proposes a JSON module for the Go
+  implementation of starlark.
+
+* [Skycfg](https://github.com/stripe/skycfg) has a `json.marshal()` method that can generate JSON documents, for use
+  with config formats based on JSON syntax.
+
+JSON implementations in the broader community have wildly variant APIs. It is likely that ad-hoc JSON extensions to
+Starlark will have different APIs between Starlark implementations.
+
+## Proposed API
+
+The JSON module is a value named `json` in the global namespace. Its API comprises the following functions. Starlark
+implementations are not required to support the entire API, but should avoid extending the API with non-standard
+functions or parameters.
+
+JSON implementations should comply with the format documented as
+[ECMA-404](https://www.ecma-international.org/publications/standards/Ecma-404.htm) and
+[RFC 8259](https://www.rfc-editor.org/rfc/rfc8259.html), which supercedes earlier drafts of JSON. In particular, the
+only permitted character encoding for contemporary JSON is UTF-8.
+
+To maintain compatibility with existing callers, new required parameters should not be added to these functions. New
+optional parameters should be defined using keyword-only parameters
+([PEP-3102](https://www.python.org/dev/peps/pep-3102/)).
+
+### json.decode()
+
+The `json.decode()` function decodes a JSON document into a Starlark value.
+
+```python
+def json.decode(data):
+```
+
+Type conversions are:
+* JSON arrays are decoded to Starlark `list` values.
+* JSON objects are decoded to Starlark `dict` values. Keys are in the same order as the input data.
+* JSON `true`, `false`, and `null` literals are decoded to Starlark `True`, `False`, and `None` respectively.
+* JSON strings are decoded to Starlark `string` values.
+* JSON numbers with no fractional component are decoded to Starlark `int` values. Starlark implementations without
+  arbitrary-precision integers should reject numbers that exceed their supported range.
+* JSON numbers with a fractional component may be decoded to an arbitrary-precision or floating-point value, if
```

Comment:
Is it conforming for implementations without floating-point values to keep them as a string? Or do they have to error?

Reply:
An implicit type conversion of float to string would be very surprising to me as a user, so I think I'd prefer the implementation to error if it can't accept a given input value.